readme.md
![gli](doc/logo.png)

[OpenGL Image](http://gli.g-truc.net/) (*GLI*) is a header only C++ image library for graphics software.

*GLI* provides classes and functions to load image files (*[KTX](https://www.khronos.org/opengles/sdk/tools/KTX/)* and *[DDS](https://msdn.microsoft.com/en-us/library/windows/desktop/bb943990%28v=vs.85%29.aspx)*), facilitate graphics APIs texture creation, compare textures, access texture texels, sample textures, convert textures, generate mipmaps, etc.

This library works perfectly with *[OpenGL](https://www.opengl.org)* or *[Vulkan](https://www.khronos.org/vulkan)* but it also ensures interoperability with other third party libraries and SDK.
It is a good candidate for software rendering (raytracing / rasterisation), image processing, image based software testing or any development context that requires a simple and convenient image library.

*GLI* is written in C++11. It is a platform independent library with no dependence and it supports the following compilers:
- [Apple Clang 5.0](https://developer.apple.com/library/mac/documentation/CompilerTools/Conceptual/LLVMCompilerOverview/index.html) and higher
- [GCC](http://gcc.gnu.org/) 4.6 and higher
- [Intel C++ Composer](https://software.intel.com/en-us/intel-compilers) XE 2013 and higher
- [LLVM](http://llvm.org/) 3.2 and higher
- [Visual C++](http://www.visualstudio.com/) 2010 and higher
- Any conform C++11 compiler

For more information about *GLI*, please have a look at the [manual](manual.md) and the [API reference documentation](http://gli.g-truc.net/0.8.0/api/index.html).
The source code and the documentation are licensed under the [Happy Bunny License (Modified MIT) or the MIT License](manual.md#section0).

Thanks for contributing to the project by [submitting issues](https://github.com/g-truc/gli/issues) for bug reports and feature requests. Any feedback is welcome at [gli@g-truc.net](mailto://gli@g-truc.net).

```cpp
#include <gli/gli.hpp>

GLuint CreateTexture(char const* Filename)
{
	gli::texture Texture = gli::load(Filename);
	if(Texture.empty())
		return 0;

	gli::gl GL(gli::gl::PROFILE_GL33);
	gli::gl::format const Format = GL.translate(Texture.format(), Texture.swizzles());
	GLenum Target = GL.translate(Texture.target());
	assert(gli::is_compressed(Texture.format()) && Target == gli::TARGET_2D);

	GLuint TextureName = 0;
	glGenTextures(1, &TextureName);
	glBindTexture(Target, TextureName);
	glTexParameteri(Target, GL_TEXTURE_BASE_LEVEL, 0);
	glTexParameteri(Target, GL_TEXTURE_MAX_LEVEL, static_cast<GLint>(Texture.levels() - 1));
	glTexParameteriv(Target, GL_TEXTURE_SWIZZLE_RGBA, &Format.Swizzles[0]);
	glTexStorage2D(Target, static_cast<GLint>(Texture.levels()), Format.Internal, Extent.x, Extent.y);
	for(std::size_t Level = 0; Level < Texture.levels(); ++Level)
	{
		glm::tvec3<GLsizei> Extent(Texture.extent(Level));
		glCompressedTexSubImage2D(
			Target, static_cast<GLint>(Level), 0, 0, Extent.x, Extent.y,
			Format.Internal, static_cast<GLsizei>(Texture.size(Level)), Texture.data(0, 0, Level));
	}

	return TextureName;
}
```

## [Lastest release](https://github.com/g-truc/gli/releases/latest)

## Project Health

| Service | System | Compiler | Status |
| ------- | ------ | -------- | ------ |
| [Drone](https://drone.io/github.com/g-truc/gli) | Linux 64 bits | GCC 4.6.3 | [![Build Status](https://drone.io/github.com/g-truc/gli/status.png)](https://drone.io/github.com/g-truc/gli/latest) |

## Release notes

### [GLI 0.8.2.0](https://github.com/g-truc/gli/releases/latest) - 201X-XX-XX
#### Features:
- Extend flip() for S3TC compressed textures #94
- Added format property queries #102

#### Improvements:
- Fixed texture operator=

---
### [GLI 0.8.1.1](https://github.com/g-truc/gli/releases/tag/0.8.1.1) - 2016-09-11
#### Improvements:
- Updated GLM to 0.9.8.0 release

#### Fixes:
- Fixed KTX cube maps saving
- Fixed texture::clear build

---
### [GLI 0.8.1.0](https://github.com/g-truc/gli/releases/tag/0.8.1.0) - 2016-03-16
#### Features:
- Added texture copy, no allocation involved, only transfer
- Added sub-image copy
- Added non-member clear
- Added make_texture* helper functions


releases.md
# Release notes

## [GLI 0.8.2.0](https://github.com/g-truc/gli/releases/latest) - 201X-XX-XX
#### Features:
- Extend flip() for S3TC compressed textures #94
- Added format property queries #102

#### Improvements:
- Fixed texture operator=

## [GLI 0.8.1.1](https://github.com/g-truc/gli/releases/tag/0.8.1.1) - 2016-09-11
#### Improvements:
- Updated GLM to 0.9.8.0 release

#### Fixes:
- Fixed KTX cube maps saving
- Fixed texture::clear build

## [GLI 0.8.1.0](https://github.com/g-truc/gli/releases/tag/0.8.1.0) - 2016-03-16
#### Features:
- Added texture copy, no allocation involved, only transfer
- Added sub-image copy
- Added non-member clear
- Added make_texture* helper functions

#### Improvements:
- Added compressed npot textures support #73
- Added image access cache to generic textures
- Added luminance alpha format translation to OpenGL 3.3+ through RG swizzling. #93
- Improved performance when sampling using mipmaps
- Improved nearest filter without border using texture_lod performance (~2.5x faster with texture2d)
- Improved texture::data() (~18x with cube array, ~68x with 2D)
- Improved texture::size() (~3.1x with cube array, ~3.9x with 2D)
- Improved simultanous texture::size() and texture::extent() calls (~2.1x with cube array, ~2.3x with 2D)

#### Fixes:
- Fixed DX10 DDS saving of 3D textures #91
- Fixed BGRX translation with PROFILE_GL33 profile #92
- Fixed DDS9 loading of 3D texture detected as 2D texture #93

## [GLI 0.8.0.0](https://github.com/g-truc/gli/releases/tag/0.8.0.0) - 2016-02-16
#### Features:
- Added texture swizzle support #79
- Added texture memory swizzle support
- Added texture conversion from any uncompressed format to any uncompressed format
- Added texture lod
- Added texture mipmaps generation for uncompressed formats
- Added support for load and store
- Added support for many new formats
- Added sampler 1D, 2D, 3D, array and cube map
- Added sampler texel fetch and texel write
- Added sampler clear
- Added transform algorithm to compute arithmetic between texels

gli/dx.hpp
///////////////////////////////////////////////////////////////////////////////////
/// OpenGL Image (gli.g-truc.net)
///
/// Copyright (c) 2008 - 2016 G-Truc Creation (www.g-truc.net)
/// Permission is hereby granted, free of charge, to any person obtaining a copy
/// of this software and associated documentation files (the "Software"), to deal
/// in the Software without restriction, including without limitation the rights
/// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
/// copies of the Software, and to permit persons to whom the Software is
/// furnished to do so, subject to the following conditions:
/// 
/// The above copyright notice and this permission notice shall be included in
/// all copies or substantial portions of the Software.
/// 
/// Restrictions:
///  By making use of the Software for military purposes, you choose to make
///  a Bunny unhappy.
/// 
/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
/// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
/// THE SOFTWARE.
///
/// @file gli/gl.hpp
/// @date 2013-11-09 / 2013-11-09
///////////////////////////////////////////////////////////////////////////////////

#pragma once

#include "format.hpp"
#include "target.hpp"
#include <array>

namespace gli
{
	/// Translation class to convert GLI enums into Direct3D enums
	class dx
	{
	public:
		#define GLI_MAKEFOURCC(ch0, ch1, ch2, ch3) \
			(std::uint32_t)( \
			(((std::uint32_t)(std::uint8_t)(ch3) << 24) & 0xFF000000) | \
			(((std::uint32_t)(std::uint8_t)(ch2) << 16) & 0x00FF0000) | \
			(((std::uint32_t)(std::uint8_t)(ch1) <<  8) & 0x0000FF00) | \
			((std::uint32_t)(std::uint8_t)(ch0)        & 0x000000FF) )

		enum d3dFormat
		{
			D3DFMT_UNKNOWN				=  0,

			D3DFMT_R8G8B8				= 20,
			D3DFMT_A8R8G8B8				= 21,
			D3DFMT_X8R8G8B8				= 22,
			D3DFMT_R5G6B5				= 23,
			D3DFMT_X1R5G5B5				= 24,
			D3DFMT_A1R5G5B5				= 25,
			D3DFMT_A4R4G4B4				= 26,
			D3DFMT_R3G3B2				= 27,
			D3DFMT_A8					= 28,
			D3DFMT_A8R3G3B2				= 29,
			D3DFMT_X4R4G4B4				= 30,
			D3DFMT_A2B10G10R10			= 31,
			D3DFMT_A8B8G8R8				= 32,
			D3DFMT_X8B8G8R8				= 33,
			D3DFMT_G16R16				= 34,
			D3DFMT_A2R10G10B10			= 35,
			D3DFMT_A16B16G16R16			= 36,

			D3DFMT_A8P8					= 40,
			D3DFMT_P8					= 41,

			D3DFMT_L8					= 50,
			D3DFMT_A8L8					= 51,
			D3DFMT_A4L4					= 52,

			D3DFMT_V8U8					= 60,
			D3DFMT_L6V5U5				= 61,
			D3DFMT_X8L8V8U8				= 62,
			D3DFMT_Q8W8V8U8				= 63,
			D3DFMT_V16U16				= 64,
			D3DFMT_A2W10V10U10			= 67,

			D3DFMT_UYVY					= GLI_MAKEFOURCC('U', 'Y', 'V', 'Y'),
			D3DFMT_R8G8_B8G8			= GLI_MAKEFOURCC('R', 'G', 'B', 'G'),
			D3DFMT_YUY2					= GLI_MAKEFOURCC('Y', 'U', 'Y', '2'),
			D3DFMT_G8R8_G8B8			= GLI_MAKEFOURCC('G', 'R', 'G', 'B'),
			D3DFMT_DXT1					= GLI_MAKEFOURCC('D', 'X', 'T', '1'),
			D3DFMT_DXT2					= GLI_MAKEFOURCC('D', 'X', 'T', '2'),
			D3DFMT_DXT3					= GLI_MAKEFOURCC('D', 'X', 'T', '3'),
			D3DFMT_DXT4					= GLI_MAKEFOURCC('D', 'X', 'T', '4'),
			D3DFMT_DXT5					= GLI_MAKEFOURCC('D', 'X', 'T', '5'),

			D3DFMT_ATI1					= GLI_MAKEFOURCC('A', 'T', 'I', '1'),
			D3DFMT_AT1N					= GLI_MAKEFOURCC('A', 'T', '1', 'N'),
			D3DFMT_ATI2					= GLI_MAKEFOURCC('A', 'T', 'I', '2'),
			D3DFMT_AT2N					= GLI_MAKEFOURCC('A', 'T', '2', 'N'),

			D3DFMT_ETC					= GLI_MAKEFOURCC('E', 'T', 'C', ' '),
			D3DFMT_ETC1					= GLI_MAKEFOURCC('E', 'T', 'C', '1'),
			D3DFMT_ATC					= GLI_MAKEFOURCC('A', 'T', 'C', ' '),
			D3DFMT_ATCA					= GLI_MAKEFOURCC('A', 'T', 'C', 'A'),
			D3DFMT_ATCI					= GLI_MAKEFOURCC('A', 'T', 'C', 'I'),

			D3DFMT_POWERVR_2BPP			= GLI_MAKEFOURCC('P', 'T', 'C', '2'),
			D3DFMT_POWERVR_4BPP			= GLI_MAKEFOURCC('P', 'T', 'C', '4'),

			D3DFMT_D16_LOCKABLE			= 70,
			D3DFMT_D32					= 71,
			D3DFMT_D15S1				= 73,
			D3DFMT_D24S8				= 75,
			D3DFMT_D24X8				= 77,
			D3DFMT_D24X4S4				= 79,
			D3DFMT_D16					= 80,

			D3DFMT_D32F_LOCKABLE		= 82,
			D3DFMT_D24FS8				= 83,

			D3DFMT_L16					= 81,

			D3DFMT_VERTEXDATA			=100,
			D3DFMT_INDEX16				=101,
			D3DFMT_INDEX32				=102,

			D3DFMT_Q16W16V16U16			=110,

			D3DFMT_MULTI2_ARGB8			= GLI_MAKEFOURCC('M','E','T','1'),

			D3DFMT_R16F					= 111,
			D3DFMT_G16R16F				= 112,
			D3DFMT_A16B16G16R16F		= 113,

			D3DFMT_R32F					= 114,
			D3DFMT_G32R32F				= 115,
			D3DFMT_A32B32G32R32F		= 116,

			D3DFMT_CxV8U8				= 117,

			D3DFMT_DX10					= GLI_MAKEFOURCC('D', 'X', '1', '0'),

			D3DFMT_GLI1					= GLI_MAKEFOURCC('G', 'L', 'I', '1'),

			D3DFMT_FORCE_DWORD			= 0x7fffffff
		};
		
		enum dxgiFormatDDS
		{
			DXGI_FORMAT_UNKNOWN							= 0,
			DXGI_FORMAT_R32G32B32A32_TYPELESS			= 1,
			DXGI_FORMAT_R32G32B32A32_FLOAT				= 2,
			DXGI_FORMAT_R32G32B32A32_UINT				= 3,
			DXGI_FORMAT_R32G32B32A32_SINT				= 4,
			DXGI_FORMAT_R32G32B32_TYPELESS				= 5,
			DXGI_FORMAT_R32G32B32_FLOAT					= 6,
			DXGI_FORMAT_R32G32B32_UINT					= 7,
			DXGI_FORMAT_R32G32B32_SINT					= 8,
			DXGI_FORMAT_R16G16B16A16_TYPELESS			= 9,
			DXGI_FORMAT_R16G16B16A16_FLOAT				= 10,
			DXGI_FORMAT_R16G16B16A16_UNORM				= 11,
			DXGI_FORMAT_R16G16B16A16_UINT				= 12,
			DXGI_FORMAT_R16G16B16A16_SNORM				= 13,
			DXGI_FORMAT_R16G16B16A16_SINT				= 14,
			DXGI_FORMAT_R32G32_TYPELESS					= 15,
			DXGI_FORMAT_R32G32_FLOAT					= 16,
			DXGI_FORMAT_R32G32_UINT						= 17,
			DXGI_FORMAT_R32G32_SINT						= 18,
			DXGI_FORMAT_R32G8X24_TYPELESS				= 19,
			DXGI_FORMAT_D32_FLOAT_S8X24_UINT			= 20,
			DXGI_FORMAT_R32_FLOAT_X8X24_TYPELESS		= 21,
			DXGI_FORMAT_X32_TYPELESS_G8X24_UINT			= 22,
			DXGI_FORMAT_R10G10B10A2_TYPELESS			= 23,
			DXGI_FORMAT_R10G10B10A2_UNORM				= 24,
			DXGI_FORMAT_R10G10B10A2_UINT				= 25,
			DXGI_FORMAT_R11G11B10_FLOAT					= 26,
			DXGI_FORMAT_R8G8B8A8_TYPELESS				= 27,
			DXGI_FORMAT_R8G8B8A8_UNORM					= 28,
			DXGI_FORMAT_R8G8B8A8_UNORM_SRGB				= 29,
			DXGI_FORMAT_R8G8B8A8_UINT					= 30,
			DXGI_FORMAT_R8G8B8A8_SNORM					= 31,
			DXGI_FORMAT_R8G8B8A8_SINT					= 32,
			DXGI_FORMAT_R16G16_TYPELESS					= 33,
			DXGI_FORMAT_R16G16_FLOAT					= 34,
			DXGI_FORMAT_R16G16_UNORM					= 35,
			DXGI_FORMAT_R16G16_UINT						= 36,
			DXGI_FORMAT_R16G16_SNORM					= 37,
			DXGI_FORMAT_R16G16_SINT						= 38,
			DXGI_FORMAT_R32_TYPELESS					= 39,
			DXGI_FORMAT_D32_FLOAT						= 40,
			DXGI_FORMAT_R32_FLOAT						= 41,
			DXGI_FORMAT_R32_UINT						= 42,
			DXGI_FORMAT_R32_SINT						= 43,
			DXGI_FORMAT_R24G8_TYPELESS					= 44,
			DXGI_FORMAT_D24_UNORM_S8_UINT				= 45,
			DXGI_FORMAT_R24_UNORM_X8_TYPELESS			= 46,
			DXGI_FORMAT_X24_TYPELESS_G8_UINT			= 47,
			DXGI_FORMAT_R8G8_TYPELESS					= 48,
			DXGI_FORMAT_R8G8_UNORM						= 49,
			DXGI_FORMAT_R8G8_UINT						= 50,
			DXGI_FORMAT_R8G8_SNORM						= 51,
			DXGI_FORMAT_R8G8_SINT						= 52,
			DXGI_FORMAT_R16_TYPELESS					= 53,
			DXGI_FORMAT_R16_FLOAT						= 54,
			DXGI_FORMAT_D16_UNORM						= 55,
			DXGI_FORMAT_R16_UNORM						= 56,
			DXGI_FORMAT_R16_UINT						= 57,
			DXGI_FORMAT_R16_SNORM						= 58,
			DXGI_FORMAT_R16_SINT						= 59,
			DXGI_FORMAT_R8_TYPELESS						= 60,
			DXGI_FORMAT_R8_UNORM						= 61,
			DXGI_FORMAT_R8_UINT							= 62,
			DXGI_FORMAT_R8_SNORM						= 63,
			DXGI_FORMAT_R8_SINT							= 64,
			DXGI_FORMAT_A8_UNORM						= 65,
			DXGI_FORMAT_R1_UNORM						= 66,
			DXGI_FORMAT_R9G9B9E5_SHAREDEXP				= 67,
			DXGI_FORMAT_R8G8_B8G8_UNORM					= 68,
			DXGI_FORMAT_G8R8_G8B8_UNORM					= 69,
			DXGI_FORMAT_BC1_TYPELESS					= 70,
			DXGI_FORMAT_BC1_UNORM						= 71,
			DXGI_FORMAT_BC1_UNORM_SRGB					= 72,
			DXGI_FORMAT_BC2_TYPELESS					= 73,
			DXGI_FORMAT_BC2_UNORM						= 74,
			DXGI_FORMAT_BC2_UNORM_SRGB					= 75,
			DXGI_FORMAT_BC3_TYPELESS					= 76,
			DXGI_FORMAT_BC3_UNORM						= 77,
			DXGI_FORMAT_BC3_UNORM_SRGB					= 78,
			DXGI_FORMAT_BC4_TYPELESS					= 79,
			DXGI_FORMAT_BC4_UNORM						= 80,
			DXGI_FORMAT_BC4_SNORM						= 81,
			DXGI_FORMAT_BC5_TYPELESS					= 82,
			DXGI_FORMAT_BC5_UNORM						= 83,
			DXGI_FORMAT_BC5_SNORM						= 84,
			DXGI_FORMAT_B5G6R5_UNORM					= 85,
			DXGI_FORMAT_B5G5R5A1_UNORM					= 86,
			DXGI_FORMAT_B8G8R8A8_UNORM					= 87,
			DXGI_FORMAT_B8G8R8X8_UNORM					= 88,
			DXGI_FORMAT_R10G10B10_XR_BIAS_A2_UNORM		= 89,
			DXGI_FORMAT_B8G8R8A8_TYPELESS				= 90,
			DXGI_FORMAT_B8G8R8A8_UNORM_SRGB				= 91,
			DXGI_FORMAT_B8G8R8X8_TYPELESS				= 92,
			DXGI_FORMAT_B8G8R8X8_UNORM_SRGB				= 93,
			DXGI_FORMAT_BC6H_TYPELESS					= 94,
			DXGI_FORMAT_BC6H_UF16						= 95,
			DXGI_FORMAT_BC6H_SF16						= 96,
			DXGI_FORMAT_BC7_TYPELESS					= 97,
			DXGI_FORMAT_BC7_UNORM						= 98,
			DXGI_FORMAT_BC7_UNORM_SRGB					= 99,
			DXGI_FORMAT_AYUV							= 100,
			DXGI_FORMAT_Y410							= 101,
			DXGI_FORMAT_Y416							= 102,
			DXGI_FORMAT_NV12							= 103,
			DXGI_FORMAT_P010							= 104,
			DXGI_FORMAT_P016							= 105,
			DXGI_FORMAT_420_OPAQUE						= 106,
			DXGI_FORMAT_YUY2							= 107,
			DXGI_FORMAT_Y210							= 108,
			DXGI_FORMAT_Y216							= 109,
			DXGI_FORMAT_NV11							= 110,
			DXGI_FORMAT_AI44							= 111,
			DXGI_FORMAT_IA44							= 112,
			DXGI_FORMAT_P8								= 113,
			DXGI_FORMAT_A8P8							= 114,
			DXGI_FORMAT_B4G4R4A4_UNORM					= 115,

			DXGI_FORMAT_P208							= 130,
			DXGI_FORMAT_V208							= 131,
			DXGI_FORMAT_V408							= 132,
			DXGI_FORMAT_ASTC_4X4_TYPELESS				= 133,
			DXGI_FORMAT_ASTC_4X4_UNORM					= 134,
			DXGI_FORMAT_ASTC_4X4_UNORM_SRGB				= 135,
			DXGI_FORMAT_ASTC_5X4_TYPELESS				= 137,
			DXGI_FORMAT_ASTC_5X4_UNORM					= 138,
			DXGI_FORMAT_ASTC_5X4_UNORM_SRGB				= 139,
			DXGI_FORMAT_ASTC_5X5_TYPELESS				= 141,
			DXGI_FORMAT_ASTC_5X5_UNORM					= 142,
			DXGI_FORMAT_ASTC_5X5_UNORM_SRGB				= 143,
			DXGI_FORMAT_ASTC_6X5_TYPELESS				= 145,
			DXGI_FORMAT_ASTC_6X5_UNORM					= 146,
			DXGI_FORMAT_ASTC_6X5_UNORM_SRGB				= 147,
			DXGI_FORMAT_ASTC_6X6_TYPELESS				= 149,
			DXGI_FORMAT_ASTC_6X6_UNORM					= 150,
			DXGI_FORMAT_ASTC_6X6_UNORM_SRGB				= 151,
			DXGI_FORMAT_ASTC_8X5_TYPELESS				= 153,
			DXGI_FORMAT_ASTC_8X5_UNORM					= 154,
			DXGI_FORMAT_ASTC_8X5_UNORM_SRGB				= 155,
			DXGI_FORMAT_ASTC_8X6_TYPELESS				= 157,
			DXGI_FORMAT_ASTC_8X6_UNORM					= 158,
			DXGI_FORMAT_ASTC_8X6_UNORM_SRGB				= 159,
			DXGI_FORMAT_ASTC_8X8_TYPELESS				= 161,
			DXGI_FORMAT_ASTC_8X8_UNORM					= 162,
			DXGI_FORMAT_ASTC_8X8_UNORM_SRGB				= 163,
			DXGI_FORMAT_ASTC_10X5_TYPELESS				= 165,
			DXGI_FORMAT_ASTC_10X5_UNORM					= 166,
			DXGI_FORMAT_ASTC_10X5_UNORM_SRGB			= 167,
			DXGI_FORMAT_ASTC_10X6_TYPELESS				= 169,
			DXGI_FORMAT_ASTC_10X6_UNORM					= 170,
			DXGI_FORMAT_ASTC_10X6_UNORM_SRGB			= 171,
			DXGI_FORMAT_ASTC_10X8_TYPELESS				= 173,
			DXGI_FORMAT_ASTC_10X8_UNORM					= 174,
			DXGI_FORMAT_ASTC_10X8_UNORM_SRGB			= 175,
			DXGI_FORMAT_ASTC_10X10_TYPELESS				= 177,
			DXGI_FORMAT_ASTC_10X10_UNORM				= 178,
			DXGI_FORMAT_ASTC_10X10_UNORM_SRGB			= 179,
			DXGI_FORMAT_ASTC_12X10_TYPELESS				= 181,
			DXGI_FORMAT_ASTC_12X10_UNORM				= 182,
			DXGI_FORMAT_ASTC_12X10_UNORM_SRGB			= 183,
			DXGI_FORMAT_ASTC_12X12_TYPELESS				= 185,
			DXGI_FORMAT_ASTC_12X12_UNORM				= 186,
			DXGI_FORMAT_ASTC_12X12_UNORM_SRGB			= 187,

			DXGI_FORMAT_FORCE_UINT						= 0xffffffffUL
		};

		enum dxgiFormatGLI
		{
			DXGI_FORMAT_R64_UINT_GLI = 1,
			DXGI_FORMAT_R64_SINT_GLI,
			DXGI_FORMAT_R64_FLOAT_GLI,
			DXGI_FORMAT_R64G64_UINT_GLI,
			DXGI_FORMAT_R64G64_SINT_GLI,
			DXGI_FORMAT_R64G64_FLOAT_GLI,
			DXGI_FORMAT_R64G64B64_UINT_GLI,
			DXGI_FORMAT_R64G64B64_SINT_GLI,
			DXGI_FORMAT_R64G64B64_FLOAT_GLI,
			DXGI_FORMAT_R64G64B64A64_UINT_GLI,
			DXGI_FORMAT_R64G64B64A64_SINT_GLI,
			DXGI_FORMAT_R64G64B64A64_FLOAT_GLI,
			
			DXGI_FORMAT_RG4_UNORM_GLI,
			DXGI_FORMAT_RGBA4_UNORM_GLI,
			DXGI_FORMAT_R5G6B5_UNORM_GLI,
			DXGI_FORMAT_R5G5B5A1_UNORM_GLI,
			DXGI_FORMAT_A1B5G5R5_UNORM_GLI,
			
			DXGI_FORMAT_R8_SRGB_GLI,
			DXGI_FORMAT_R8_USCALED_GLI,
			DXGI_FORMAT_R8_SSCALED_GLI,
			
			DXGI_FORMAT_R8G8_SRGB_GLI,
			DXGI_FORMAT_R8G8_USCALED_GLI,
			DXGI_FORMAT_R8G8_SSCALED_GLI,
			
			DXGI_FORMAT_R8G8B8_UNORM_GLI,
			DXGI_FORMAT_R8G8B8_SNORM_GLI,
			DXGI_FORMAT_R8G8B8_USCALED_GLI,
			DXGI_FORMAT_R8G8B8_SSCALED_GLI,
			DXGI_FORMAT_R8G8B8_UINT_GLI,
			DXGI_FORMAT_R8G8B8_SINT_GLI,
			DXGI_FORMAT_R8G8B8_SRGB_GLI,
			
			DXGI_FORMAT_B8G8R8_UNORM_GLI,
			DXGI_FORMAT_B8G8R8_SNORM_GLI,
			DXGI_FORMAT_B8G8R8_USCALED_GLI,
			DXGI_FORMAT_B8G8R8_SSCALED_GLI,
			DXGI_FORMAT_B8G8R8_UINT_GLI,
			DXGI_FORMAT_B8G8R8_SINT_GLI,
			DXGI_FORMAT_B8G8R8_SRGB_GLI,
			
			DXGI_FORMAT_R8G8B8A8_USCALED_GLI,
			DXGI_FORMAT_R8G8B8A8_SSCALED_GLI,
			
			DXGI_FORMAT_B8G8R8A8_SNORM_GLI,
			DXGI_FORMAT_B8G8R8A8_USCALED_GLI,
			DXGI_FORMAT_B8G8R8A8_SSCALED_GLI,
			DXGI_FORMAT_B8G8R8A8_UINT_GLI,
			DXGI_FORMAT_B8G8R8A8_SINT_GLI,
			
			DXGI_FORMAT_R8G8B8A8_PACK_UNORM_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_SNORM_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_USCALED_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_SSCALED_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_UINT_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_SINT_GLI,
			DXGI_FORMAT_R8G8B8A8_PACK_SRGB_GLI,
			
			DXGI_FORMAT_R10G10B10A2_SNORM_GLI,
			DXGI_FORMAT_R10G10B10A2_USCALED_GLI,
			DXGI_FORMAT_R10G10B10A2_SSCALED_GLI,
			DXGI_FORMAT_R10G10B10A2_SINT_GLI,
			
			DXGI_FORMAT_B10G10R10A2_UNORM_GLI,
			DXGI_FORMAT_B10G10R10A2_SNORM_GLI,
			DXGI_FORMAT_B10G10R10A2_USCALED_GLI,
			DXGI_FORMAT_B10G10R10A2_SSCALED_GLI,
			DXGI_FORMAT_B10G10R10A2_UINT_GLI,
			DXGI_FORMAT_B10G10R10A2_SINT_GLI,
			
			DXGI_FORMAT_R16_USCALED_GLI,
			DXGI_FORMAT_R16_SSCALED_GLI,
			DXGI_FORMAT_R16G16_USCALED_GLI,
			DXGI_FORMAT_R16G16_SSCALED_GLI,
			
			DXGI_FORMAT_R16G16B16_UNORM_GLI,
			DXGI_FORMAT_R16G16B16_SNORM_GLI,
			DXGI_FORMAT_R16G16B16_USCALED_GLI,
			DXGI_FORMAT_R16G16B16_SSCALED_GLI,
			DXGI_FORMAT_R16G16B16_UINT_GLI,
			DXGI_FORMAT_R16G16B16_SINT_GLI,
			DXGI_FORMAT_R16G16B16_FLOAT_GLI,
			
			DXGI_FORMAT_R16G16B16A16_USCALED_GLI,
			DXGI_FORMAT_R16G16B16A16_SSCALED_GLI,
			
			DXGI_FORMAT_S8_UINT_GLI,
			DXGI_FORMAT_D16_UNORM_S8_UINT_GLI,
			DXGI_FORMAT_D24_UNORM_GLI,
			
			DXGI_FORMAT_L8_UNORM_GLI,
			DXGI_FORMAT_A8_UNORM_GLI,
			DXGI_FORMAT_LA8_UNORM_GLI,
			DXGI_FORMAT_L16_UNORM_GLI,
			DXGI_FORMAT_A16_UNORM_GLI,
			DXGI_FORMAT_LA16_UNORM_GLI,
			
			DXGI_FORMAT_R3G3B2_UNORM_GLI,
			
			DXGI_FORMAT_BC1_RGB_UNORM_GLI,
			DXGI_FORMAT_BC1_RGB_SRGB_GLI,
			DXGI_FORMAT_RGB_ETC2_UNORM_GLI,
			DXGI_FORMAT_RGB_ETC2_SRGB_GLI,
			DXGI_FORMAT_RGBA_ETC2_A1_UNORM_GLI,
			DXGI_FORMAT_RGBA_ETC2_A1_SRGB_GLI,
			DXGI_FORMAT_RGBA_ETC2_UNORM_GLI,
			DXGI_FORMAT_RGBA_ETC2_SRGB_GLI,
			DXGI_FORMAT_R11_EAC_UNORM_GLI,
			DXGI_FORMAT_R11_EAC_SNORM_GLI,
			DXGI_FORMAT_RG11_EAC_UNORM_GLI,
			DXGI_FORMAT_RG11_EAC_SNORM_GLI,
			
			DXGI_FORMAT_RGB_PVRTC1_8X8_UNORM_GLI,
			DXGI_FORMAT_RGB_PVRTC1_8X8_SRGB_GLI,
			DXGI_FORMAT_RGB_PVRTC1_16X8_UNORM_GLI,
			DXGI_FORMAT_RGB_PVRTC1_16X8_SRGB_GLI,
			DXGI_FORMAT_RGBA_PVRTC1_8X8_UNORM_GLI,
			DXGI_FORMAT_RGBA_PVRTC1_8X8_SRGB_GLI,
			DXGI_FORMAT_RGBA_PVRTC1_16X8_UNORM_GLI,
			DXGI_FORMAT_RGBA_PVRTC1_16X8_SRGB_GLI,
			DXGI_FORMAT_RGBA_PVRTC2_8X8_UNORM_GLI,
			DXGI_FORMAT_RGBA_PVRTC2_8X8_SRGB_GLI,
			DXGI_FORMAT_RGBA_PVRTC2_16X8_UNORM_GLI,
			DXGI_FORMAT_RGBA_PVRTC2_16X8_SRGB_GLI,
			
			DXGI_FORMAT_RGB_ETC_UNORM_GLI,
			DXGI_FORMAT_RGB_ATC_UNORM_GLI,
			DXGI_FORMAT_RGBA_ATCA_UNORM_GLI,
			DXGI_FORMAT_RGBA_ATCI_UNORM_GLI,
		};
		
		union dxgiFormat
		{
			dxgiFormat()
				: DDS(DXGI_FORMAT_UNKNOWN)
			{}
			
			dxgiFormat(dxgiFormatDDS DDS)
				: DDS(DDS)
			{}

			dxgiFormat(dxgiFormatGLI GLI)
				: GLI(GLI)
			{}
			
			dxgiFormatDDS DDS;
			dxgiFormatGLI GLI;
		};
		
		enum ddpf
		{
			DDPF_ALPHAPIXELS = 0x1,
			DDPF_ALPHA = 0x2,
			DDPF_FOURCC = 0x4,
			DDPF_RGB = 0x40,
			DDPF_YUV = 0x200,
			DDPF_LUMINANCE = 0x20000,
			DDPF_LUMINANCE_ALPHA = DDPF_LUMINANCE | DDPF_ALPHA,
			DDPF_RGBAPIXELS = DDPF_RGB | DDPF_ALPHAPIXELS,
			DDPF_RGBA = DDPF_RGB | DDPF_ALPHA,
			DDPF_LUMINANCE_ALPHAPIXELS = DDPF_LUMINANCE | DDPF_ALPHAPIXELS,

		};

		struct format
		{
			ddpf DDPixelFormat;
			d3dFormat D3DFormat;
			dxgiFormat DXGIFormat;
			glm::u32vec4 Mask;
		};

	public:
		dx();

		/// Convert GLI formats into Direct3D formats
		format const & translate(gli::format const & Format) const;

		/// Convert a Direct3D 9 format into a GLI format
		gli::format find(d3dFormat FourCC, ddpf PixelFormat) const;

		/// Convert a Direct3D 10 format into a GLI format
		gli::format find(d3dFormat FourCC, dxgiFormat Format, ddpf PixelFormat) const;

	private:
		std::array<format, FORMAT_COUNT> Translation;
	};

	/// Evaluate whether a target and format combinaison is only supported by the DDS container through GLI DDS extension.
	bool is_dds_ext(target Target, format Format);
}//namespace gli

#include "./core/dx.inl"
